Part 1: Cross Navigation Between NavHosts

skaik-mo opened the pull request.

dev-bilal-azzam committed:
try navigation using API

dev-bilal-azzam committed:
implement details api

dev-bilal-azzam committed:
implement intra details navigation

dev-bilal-azzam committed:
implement explore api

dev-bilal-azzam committed:
implement intra explore feature navigation

dev-bilal-azzam committed:
implement cross feature navigation

dev-bilal-azzam committed:
movie api implementation files to the higher level package of the feature

dev-bilal-azzam committed:
Merge branch 'feature/main-nav' into feature/cineverse-cross-feature-nav

# Conflicts:
#	api/details/build.gradle.kts
#	api/details/src/main/java/com/giraffe/details/DetailsApi.kt
#	api/explore/build.gradle.kts
#	api/explore/src/main/java/com/giraffe/explore/ExploreApi.kt
#	app/src/main/java/com/giraffe/cineverseapp/MainActivity.kt
#	app/src/main/java/com/giraffe/cineverseapp/di/featureApiModule.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/moviedetails/screen/MovieDetailsRoute.kt
#	presentation/explore/src/main/java/com/giraffe/explore/components/TransitionLazyColumnToGrid.kt
#	presentation/explore/src/main/java/com/giraffe/explore/nav/route/DiscoverRoute.kt
#	presentation/explore/src/main/java/com/giraffe/explore/nav/route/SearchResultRoute.kt
#	presentation/explore/src/main/java/com/giraffe/explore/screen/discover/DiscoverScreen.kt
#	presentation/explore/src/main/java/com/giraffe/explore/screen/searchresult/SearchResultScreen.kt

dev-bilal-azzam committed:
remove duplicate , temp and unnedded classes

dev-bilal-azzam committed:
edit screens and routes to adopt navigation changes

dev-bilal-azzam committed:
clean up cross feature navigation

dev-bilal-azzam committed:
clean up unused imports

dev-bilal-azzam committed:
clean up unused imports

skaik-mo committed:
Refactor DetailsApi and NavGraph

- Move DetailsApiImp to nav package
- Create DetailsNavGraph composable
- Add onBackButtonClick to movieDetailsRoute

Mostafa-khaled21 commented on presentation/details/src/main/java/com/giraffe/details/screens/recommended/RecommendedEffect.kt:
why did you delete recommended package ? we are using it.

skaik-mo replied:
I think Bilal just removed it while he was testing, and we will bring it back from develop.

dev-bilal-azzam committed:
Merge remote-tracking branch 'origin/feature/cineverse-cross-feature-nav' into feature/cineverse-cross-feature-nav

skaik-mo committed:
Refactor: Migrate Explore module to use type-safe navigation

Key changes:
- Moved `ExploreApiImp.kt` to the `navigation` package.
- Updated `ExploreApiImp` to use the new `ExploreNavGraph`.
- Replaced string-based navigation routes with serializable data objects for `DiscoverRoute`, `SearchRoute`, and `SearchResultRoute`.
- Updated `ExploreNavGraph` to use the new type-safe routes and pass necessary navigation actions and callbacks.
- Adjusted screen composables (`DiscoverScreen`, `SearchScreen`, `SearchResultScreen`) to accept navigation callbacks directly instead of `NavController`.
- Updated the `featureApiModule` to provide the refactored `ExploreApiImp`.

K-M-9 commented on api/explore/src/main/java/com/giraffe/explore/ExploreApi.kt:
```diff
@@ -1,23 +1,13 @@
-    fun NavGraphBuilder.searchResultGraph(
-        navController: NavHostController,
+    @Composable
+    fun ExploreContainer(
+        modifier: Modifier,
```
Don’t pass `Compose’s Modifier` into your `API` interface `function` doing so `couples` your interface to `Compose` and will cause `problems` if you later `switch` to a different `implementation`.

dev-bilal-azzam replied:
The Explore Container is already a composable.
So if we decided to switch to different implementation the change will happen in the implementation not the API.

K-M-9 replied:
If you switch the implementation to `XML` navigation, you won’t need `Modifier` because `XML` doesn’t use it.

dev-bilal-azzam replied:
If you're using compose to build ui then decided to switch specific portion of the ui (feature) into XML view, you just wrap your new XML layout inside a composable AndroidView which also needs a modifier.

- In general as long as your app using compose it have to send a modifier for each screen.

K-M-9 replied:
Even though your app uses Compose and often requires passing a `Modifier`, that doesn’t mean your `API` interfaces should be aware of it. Interfaces should stay `independent` of the `UI` framework to allow `flexible` implementation changes, like `switching` from.

dev-bilal-azzam replied:
Using modifiers here is essential as long as the app is using compose "the app not the feature" and also it doesn't hinder or prevent the feature from switching to another implementation at all.
and the interface will depend on compose as long as the app is using compose.

K-M-9 replied:
Even if the app uses `Compose`, that doesn't justify leaking Compose specific `types` like `Modifier` into interfaces, clean separation keeps your architecture flexible and decoupled from specific `UI` frameworks.

dev-bilal-azzam replied:
How that comes the interface is already using @composable so it is already depending on compose and this doesn't hinder at all switching ui framework in a specific feature. if we migrated a specific feature to xml it will also be wrapped in compose so it will still needs modifier.

- I don't see any point here, Did you tried to use partially xml layouts in compose?

K-M-9 replied:
Marking a function as `@Composable` ties it to Compose at `runtime`, but passing `Modifier` leaks layout concerns into your interface.
A `Composable` interface can still be UI agnostic `unlike` `Modifier`, which `tightly` couples it to Compose’s `UI` system.

dev-bilal-azzam replied:
if I'd call a composable function I have to send a modifier with it to control its layout and I still can't see wht the problem this would result. So can you give an example of how sending the Modifier conflicts with changing the implementation?

K-M-9 replied:
You’ve made the `API` interface aware of `Compose` specific `types`, which becomes a problem if you switch to `XML` since `XML` doesn’t use `Modifier`.
An `API` interface shouldn’t know anything about the underlying UI implementation.

K-M-9 commented on api/details/src/main/java/com/giraffe/details/DetailsApi.kt:
```diff
@@ -1,24 +1,14 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
 
+// the api represent cross feature interaction
```
remove the `comment`.

K-M-9 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
+
```
Remove the `extra` spaces in the file and `unstage` it from your changes.

K-M-9 commented on app/src/main/java/com/giraffe/cineverseapp/di/viewModelModule.kt:
```diff
@@ -18,9 +15,9 @@ val viewModelModule = module {
     viewModelOf(::SearchResultViewModel)
-    viewModelOf(::RecommendedSeriesViewModel)
-    viewModel { (seriesId: Long) -> SeriesRecommendationViewModel(get(), seriesId) }
+    //viewModelOf(::RecommendedSeriesViewModel)
+    //viewModel { (seriesId: Long) -> SeriesRecommendationViewModel(get(), seriesId) }
```
You’ve commented out those `viewModelOf` calls, if they’re not `needed`, simply `delete` those lines.

dev-bilal-azzam replied:
They are needed but will be changed soon

K-M-9 commented on app/src/main/java/com/giraffe/cineverseapp/MainActivity.kt:
```diff
@@ -4,20 +4,29 @@ import android.os.Bundle
-                detailsApi = DetailsApiImp()
-            )
+
+            val detailsApi: DetailsApi by inject()
+            val exploreApi: ExploreApi by inject()
```
Move the definitions of `detailsApi: DetailsApi by inject()` and `exploreApi: ExploreApi by inject()` out of the `setContent` and `onCreate` block.

dev-bilal-azzam replied:
done

K-M-9 commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsRoute.kt:
```diff
@@ -1,8 +1,32 @@
+            navArgument(CAST_ID_ARG) {
+                type = NavType.IntType
+            })) { backStackEntry ->
+        val personId = backStackEntry.arguments?.getInt(CAST_ID_ARG) ?: 268
```
Don’t fall back to a hardcoded default for `personId`. If the argument is missing, the ViewModel should handle the error ,for example, by showing an error message or screen instead of `assuming` a specific ID.

dev-bilal-azzam replied:
done

skaik-mo replied:
We rehandle any things related to details route in other PR to make other PR it simple and small files

mohannadahmed00 commented on app/src/main/java/com/giraffe/cineverseapp/nav/CineverseNavigator.kt:
rename this file to AppNavigator to be easier to access and read.

dev-bilal-azzam replied:
done

mohannadahmed00 commented on app/src/main/java/com/giraffe/cineverseapp/nav/featurescreen.kt:
rename this file to AppScreen to be easier to access and read.

mohannadahmed00 commented on domain/media/src/main/java/com/giraffe/media/movies/usecase/GetMoviesGenresUseCase.kt:
```diff
@@ -7,6 +7,7 @@ class GetMoviesGenresUseCase(
     private val repository: MoviesRepository
 ) {
     suspend operator fun invoke(): List<Genre> {
+        println("GetMoviesGenresUseCase")
```
remove this println.

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/nav/DetailsApiImp.kt:
```diff
@@ -0,0 +1,35 @@
+            is DetailsStartDestination.Series -> "${SERIES_ROUTE}/${startDestination.seriesId}"
+            is DetailsStartDestination.Cast -> "${CAST_ROUTE}/${startDestination.castId}"
+        }
+        Log.d("TAG DetailsContainer:", "$startDestinationRoute")
```
remove this log

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/nav/DetailsNavGraph.kt:
```diff
@@ -0,0 +1,58 @@
+        movieDetailsRoute(
+            navController = navController,
+            onBackButtonClick = {
+                val istt = navController.popBackStack()
```
remove it and the log if you it isn't needed

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsScreen.kt:
```diff
@@ -67,7 +66,7 @@ fun CastDetailsScreen(
             is CastDetailsEffect.NavigateToMovies -> {}
             is CastDetailsEffect.NavigateToGallery -> {
-                navController.navigateToGallery(it.actorName, it.imageUrls)
+                //navController.navigateToGallery(it.actorName, it.imageUrls)
```
remove it if it's not needed

skaik-mo replied:
we rehandle any things related to details route in other PR to make other PR it simple and small files

dev-bilal-azzam replied:
needed for coming changes on another PR

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/gallery/Navigator.kt:
remove this file

dev-bilal-azzam replied:
will be needed for coming changes in another PR

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/seasons/SeasonRoute.kt:
add its route or remove it if it's not needed

dev-bilal-azzam replied:
needed for coming changes on another PR

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/seriesdetails/SeriesDetailsScreen.kt:
```diff
@@ -67,13 +64,15 @@ fun SeriesDetailsScreen(
-                seriesId = it.seriesId
-            )
+            is SeriesDetailsEffect.NavigateToSeasons -> {
+                /*navController.navigateToSeasons(
```
remove this commented code

dev-bilal-azzam replied:
will be needed for coming changes in another PR

mohannadahmed00 commented on presentation/details/src/main/java/com/giraffe/details/screens/seriesdetails/SeriesDetailsScreen.kt:
```diff
@@ -194,14 +193,14 @@ fun SeriesDetailsContent(
                 movies = state.recommendedSeries,
                 onClickEndText = {
-                    navController.navigate(
+                    /*navController.navigateT(
```
remove this commented code

dev-bilal-azzam replied:
needed for coming changes on another PR

mohannadahmed00 commented on presentation/explore/src/main/java/com/giraffe/explore/screen/discover/DiscoverViewModel.kt:
```diff
@@ -15,14 +16,18 @@ class DiscoverViewModel(
 
+    private val TAG = "TAG"
     init {
+        Log.d(TAG, "discover vm: init")
```
remove all logs in this file

dev-bilal-azzam replied:
done

mohannadahmed00 commented on repository/media/src/main/java/com/giraffe/media/utils/SafeCall.kt:
```diff
@@ -35,9 +35,10 @@ import java.net.UnknownHostException
     suspend operator fun <T> invoke(execute: suspend () -> T): T {
         return try {
-            Log.d("genresExecute", "loadExecute: ${execute.invoke()}")
+            Log.d("TAG", "invoke: ")
```
remove this log

dev-bilal-azzam replied:
done

K-M-9 commented:
There are multiple `log` statements and `comments`, and it would be better to use `Serializable` instead of use `strings`.

dev-bilal-azzam committed:
rename app navigator and app screen

dev-bilal-azzam committed:
handle sending invalid cast id instead of default value

dev-bilal-azzam committed:
movie injecting apis outside on create

dev-bilal-azzam committed:
handle back buttons click

dev-bilal-azzam committed:
remove debug code logs

dev-bilal-azzam committed:
remove unneeded comments

dev-bilal-azzam committed:
handle cross-feature navigation state by creating back stack

dev-bilal-azzam committed:
optimize imports

dev-bilal-azzam committed:
Merge branch 'feature/cineverse-nav-back-stack' into feature/cineverse-cross-feature-nav

# Conflicts:
#	app/src/main/java/com/giraffe/cineverseapp/nav/AppNavigator.kt

dev-bilal-azzam committed:
remove modifier from feature api to make each feature independent

K-M-9 commented on presentation/details/src/main/java/com/giraffe/details/screens/moviedetails/screen/MovieDetailsRoute.kt:
```diff
@@ -1,8 +1,37 @@
+                type = NavType.IntType
+            })
+    ) { backStackEntry ->
+        val movieID = backStackEntry.arguments?.getInt(MOVIE_ID_ARG) ?: 268
```
Don’t fall back to a hardcoded default for `movieID`. If the argument is missing, the `ViewModel` should handle the error ,for example, by showing an error message or screen instead of assuming a specific `ID`.
